File: calc/javacalc/src/main/java/com/javacalc/CalculatorModel.java
package com.javacalc;

public class CalculatorModel {
    private double storedValue = 0;
    private String currentInput = "0";
    private Operation currentOperator = null;
    private boolean startNewNumber = true;
    private boolean isError = false;
    private String expressionText = " ";

    public CalculatorModel() {
        reset();
    }

    public void reset() {
        storedValue = 0;
        currentInput = "0";
        currentOperator = null;
        startNewNumber = true;
        isError = false;
        expressionText = " ";
    }

    public void clearEntry() {
        currentInput = "0";
        startNewNumber = true;
        isError = false;
    }

    public void enterDigit(String digit) {
        if (startNewNumber || isError) {
            currentInput = digit;
            startNewNumber = false;
            isError = false;
        } else {
            if (currentInput.equals("0")) {
                currentInput = digit;
            } else {
                currentInput += digit;
            }
        }
    }

    public void enterDecimal() {
        if (startNewNumber || isError) {
            currentInput = "0.";
            startNewNumber = false;
            isError = false;
        } else if (!currentInput.contains(".")) {
            currentInput += ".";
        }
    }

    public void backspace() {
        if (startNewNumber || isError) return;

        if (currentInput.length() > 1) {
            currentInput = currentInput.substring(0, currentInput.length() - 1);
            if (currentInput.equals("-")) {
                currentInput = "0";
                startNewNumber = true;
            }
        } else {
            currentInput = "0";
            startNewNumber = true;
        }
    }

    public void toggleSign() {
        if (isError) return;
        try {
            double val = Double.parseDouble(currentInput);
            if (val != 0) {
                val = -val;
                currentInput = formatNumber(val);
            }
        } catch (NumberFormatException ignored) {}
    }

    public void setOperator(Operation op) {
        if (isError) return;

        try {
            double currentValue = Double.parseDouble(currentInput);

            if (currentOperator != null && !startNewNumber) {
                currentValue = currentOperator.apply(storedValue, currentValue);
                currentInput = formatNumber(currentValue);
            }

            storedValue = currentValue;
            currentOperator = op;
            expressionText = formatNumber(storedValue) + " " + op.getDisplaySymbol();
            startNewNumber = true;
        } catch (ArithmeticException ex) {
            handleError(ex.getMessage());
        } catch (NumberFormatException ignored) {}
    }

    public void calculateEquals() {
        if (isError || currentOperator == null) return;

        try {
            double secondOperand = Double.parseDouble(currentInput);
            expressionText = formatNumber(storedValue) + " " + currentOperator.getDisplaySymbol() + " " + formatNumber(secondOperand) + " =";

            double result = currentOperator.apply(storedValue, secondOperand);
            currentInput = formatNumber(result);

            storedValue = result;
            currentOperator = null;
            startNewNumber = true;
        } catch (ArithmeticException ex) {
            handleError("Error: Division by zero");
        } catch (NumberFormatException ignored) {}
    }

    private void handleError(String message) {
        currentInput = message;
        expressionText = " ";
        isError = true;
        startNewNumber = true;
    }

    public String formatNumber(double value) {
        if (Double.isNaN(value) || Double.isInfinite(value)) {
            return "Error";
        }
        if (value == (long) value) {
            return String.format("%d", (long) value);
        }
        return String.valueOf(value);
    }

    // Getters
    public String getCurrentInput() {
        return currentInput;
    }

    public String getExpressionText() {
        return expressionText;
    }

    public boolean isError() {
        return isError;
    }

    public Operation getCurrentOperator() {
        return currentOperator;
    }
}

File: calc/fluttercalc/lib/calculator_model.dart
import 'operation.dart';

class DivisionByZeroException implements Exception {
  final String message;
  DivisionByZeroException(this.message);
}

class CalculatorModel {
  double _storedValue = 0;
  String _currentInput = '0';
  Operation? _currentOperator;
  bool _startNewNumber = true;
  bool _isError = false;
  String _expressionText = ' ';

  CalculatorModel() {
    reset();
  }

  void reset() {
    _storedValue = 0;
    _currentInput = '0';
    _currentOperator = null;
    _startNewNumber = true;
    _isError = false;
    _expressionText = ' ';
  }

  void clearEntry() {
    _currentInput = '0';
    _startNewNumber = true;
    _isError = false;
  }

  void enterDigit(String digit) {
    if (_startNewNumber || _isError) {
      _currentInput = digit;
      _startNewNumber = false;
      _isError = false;
    } else if (_currentInput == '0') {
      _currentInput = digit;
    } else {
      _currentInput += digit;
    }
  }

  void enterDecimal() {
    if (_startNewNumber || _isError) {
      _currentInput = '0.';
      _startNewNumber = false;
      _isError = false;
    } else if (!_currentInput.contains('.')) {
      _currentInput += '.';
    }
  }

  void backspace() {
    if (_startNewNumber || _isError) return;

    if (_currentInput.length > 1) {
      _currentInput = _currentInput.substring(0, _currentInput.length - 1);
      if (_currentInput == '-') {
        _currentInput = '0';
        _startNewNumber = true;
      }
    } else {
      _currentInput = '0';
      _startNewNumber = true;
    }
  }

  void toggleSign() {
    if (_isError) return;
    final val = double.tryParse(_currentInput);
    if (val == null) return;
    if (val != 0) {
      _currentInput = formatNumber(-val);
    }
  }

  void setOperator(Operation op) {
    if (_isError) return;

    final currentValue = double.tryParse(_currentInput);
    if (currentValue == null) return;
    var newValue = currentValue;

    try {
      if (_currentOperator != null && !_startNewNumber) {
        newValue = _currentOperator!.apply(_storedValue, currentValue);
        _currentInput = formatNumber(newValue);
      }

      _storedValue = newValue;
      _currentOperator = op;
      _expressionText = '${formatNumber(_storedValue)} ${op.displaySymbol}';
      _startNewNumber = true;
    } on ArgumentError catch (ex) {
      _handleError(ex.message.toString());
    }
  }

  void calculateEquals() {
    if (_isError || _currentOperator == null) return;

    final secondOperand = double.tryParse(_currentInput);
    if (secondOperand == null) return;

    try {
      _expressionText =
          '${formatNumber(_storedValue)} ${_currentOperator!.displaySymbol} ${formatNumber(secondOperand)} =';

      final result = _currentOperator!.apply(_storedValue, secondOperand);
      _currentInput = formatNumber(result);

      _storedValue = result;
      _currentOperator = null;
      _startNewNumber = true;
    } on ArgumentError {
      _handleError('Error: Division by zero');
    }
  }

  void _handleError(String message) {
    _currentInput = message;
    _expressionText = ' ';
    _isError = true;
    _startNewNumber = true;
  }

  String formatNumber(double value) {
    if (value.isNaN || value.isInfinite) {
      return 'Error';
    }
    if (value == value.truncateToDouble()) {
      return value.truncate().toString();
    }
    return value.toString();
  }

  // Getters
  String get currentInput => _currentInput;
  String get expressionText => _expressionText;
  bool get isError => _isError;
  Operation? get currentOperator => _currentOperator;
}
